server/data/data-jmap-postgres/src/main/java/org/apache/james/jmap/postgres/projections/PostgresEmailQueryViewDAO.java

```diff
@@ -0,0 +1,145 @@
+
+        return postgresExecutor.executeRows(dslContext -> Flux.from(dslContext.select(MESSAGE_ID)
+                .from(TABLE_NAME)
+                .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+                .and(RECEIVED_AT.greaterOrEqual(since.toOffsetDateTime()))
+                .orderBy(RECEIVED_AT.desc())
+                .limit(limit.getLimit().get())))
+            .map(record -> PostgresMessageId.Factory.of(record.get(MESSAGE_ID)));
+    }
+
+    public Flux<MessageId> listMailboxContentBeforeSortedByReceivedAt(PostgresMailboxId mailboxId, ZonedDateTime since, Limit limit) {
+        Preconditions.checkArgument(!limit.isUnlimited(), "Limit should be defined");
+
+        return postgresExecutor.executeRows(dslContext -> Flux.from(dslContext.select(MESSAGE_ID)
+                .from(TABLE_NAME)
+                .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+                .and(RECEIVED_AT.lessOrEqual(since.toOffsetDateTime()))
+                .orderBy(RECEIVED_AT.desc())
+                .limit(limit.getLimit().get())))
+            .map(record -> PostgresMessageId.Factory.of(record.get(MESSAGE_ID)));
+    }
+
+    public Flux<MessageId> listMailboxContentSinceSentAt(PostgresMailboxId mailboxId, ZonedDateTime since, Limit limit) {
+        Preconditions.checkArgument(!limit.isUnlimited(), "Limit should be defined");
+
+        return postgresExecutor.executeRows(dslContext -> Flux.from(dslContext.select(MESSAGE_ID)
+                .from(TABLE_NAME)
+                .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+                .and(SENT_AT.greaterOrEqual(since.toOffsetDateTime()))
+                .orderBy(SENT_AT.desc())
+                .limit(limit.getLimit().get())))
+            .map(record -> PostgresMessageId.Factory.of(record.get(MESSAGE_ID)));
+    }
+
+    public Mono<Void> delete(PostgresMailboxId mailboxId, PostgresMessageId messageId) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
+            .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+            .and(MESSAGE_ID.eq(messageId.asUuid()))));
+    }
+
+    public Mono<Void> delete(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
+            .where(MAILBOX_ID.eq(mailboxId.asUuid()))));
+    }
+
+    public Mono<Void> save(PostgresMailboxId mailboxId, ZonedDateTime sentAt, ZonedDateTime receivedAt, PostgresMessageId messageId) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.insertInto(TABLE_NAME)
+            .set(MAILBOX_ID, mailboxId.asUuid())
+            .set(MESSAGE_ID, messageId.asUuid())
+            .set(SENT_AT, sentAt.toOffsetDateTime())
+            .set(RECEIVED_AT, receivedAt.toOffsetDateTime())
+            .onConflictOnConstraint(PK_CONSTRAINT_NAME)
+            .doUpdate()
+            .set(SENT_AT, sentAt.toOffsetDateTime())
+            .set(RECEIVED_AT, receivedAt.toOffsetDateTime())));
```

Comment:
This is supposed to be indempotent so I bet in case of conflict we can just do nothing?

Reply:
actually in the contract, there is a test:

```
@Test
    default void saveShouldBeIdempotent() {
        testee().save(mailboxId1(), DATE_1, DATE_2, messageId1()).block();
        testee().save(mailboxId1(), DATE_1, DATE_2, messageId1()).block();

        assertThat(testee().listMailboxContentSortedBySentAt(mailboxId1(), Limit.limit(12)).collectList().block())
            .containsExactly(messageId1());
    }
```

Does this mean in case of conflict we just do nothing?

Reply:
Idempotent means that if you insert the same record twice you still have the same result. The sent_at and received_at here are not supposed to change, thus why you can simply just do nothing, result will be the same, as for the test.